### PR TITLE
New AESM protobuf APIs

### PR DESCRIPTION
**Pull request by AdrianCX**
Changes needed for: https://github.com/fortanix/rust-sgx/issues/231

### CODE REVIEW

File: aesm-client/Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ test-sgx = []
 # Project dependencies
 sgxs = { version = "0.7.0", path = "../sgxs", optional = true }
 sgx-isa = { version = "0.3.0", path = "../sgx-isa"}
+sgx_types = { version = "1.1.1" }
```
Comment (jethrogb) [inline aesm-client/Cargo.toml:31]: This is in the wrong section. This is an external dependency.

File: aesm-client/tests/live_quote.rs [hunk 2]
```diff
@@ -41,3 +44,38 @@ fn live_quote() {
         )
         .expect("quote result");
 }
+
+#[test]
+fn test_ecdsa_quote() {
+    // Doing a meaningful test of this requires creating an enclave, this is
+    // just a simple test that we can send a bogus request and get an error
+    // back. The node attest flow in testsetup.sh exercises the real case.
```
Comment (jethrogb) [inline aesm-client/tests/live_quote.rs:52]: Uh, what now?

File: aesm-client/tests/live_quote.rs [hunk 3]
```diff
@@ -41,3 +44,38 @@ fn live_quote() {
...
+    const DUMMY_SPID: [u8; 16] = [0; 16];
```
Comment (jethrogb) [inline aesm-client/tests/live_quote.rs:53]: Make sure to fix all warnings

File: dcap-provider/src/lib.rs [hunk 4]
```diff
@@ -262,9 +262,12 @@ pub extern "C" fn sgx_ql_get_quote_config(
                         // variable, but there is no guarantee that the second
                         // call to this function is going to happen on the same
                         // stack.
+
+                        // We may get called back from aesm service if library injected there.
+                        // In that case we need to return PlatformLibUnavailable in order to have AESM use internal data.
                         let _guard = ENTERED_ONCE
                             .try_lock()
-                            .map_err(|_| Quote3Error::NoPlatformCertData)?;
+                            .map_err(|_| Quote3Error::PlatformLibUnavailable)?;
```
Comment (jethrogb) [inline dcap-provider/src/lib.rs:270]: This should not be in this PR

File: aesm-client/src/lib.rs [hunk 5]
```diff
@@ -218,6 +265,27 @@ impl AesmClient {
             attributes,
         )
     }
+
+    pub fn init_quote_ecdsa_p256(&self) -> Result<QuoteInfoEx> {
```
Comment (jethrogb) [inline aesm-client/src/lib.rs:269]: Remove this wrapper, you should expose `get_supported_att_key_ids` and `init_quote_ex` directly

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 6]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+        let res = self.transact(req)?;
+
+        Ok(res.get_att_key_id_num())
+    }
+
+    ///
+    /// Returns all keys supported by AESM service.
+    /// Similar functionality to sgx_get_supported_att_key_ids in page 177 at https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Developer_Reference_Linux_2.9.1_Open_Source.pdf
+    ///
+    /// @returns structure containing set of keys that can serve as input for init_quote_ex/get_quote_size_ex operations. (see AesmKeyIds::select_algorithm_id to extract a specific key)
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:136]: Please don't use javadoc syntax in Rust documentation

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 7]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+    ///
+    pub fn get_supported_att_key_ids(&self) -> Result<AesmKeyIds> {
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:138]: I think we should return a Vec<Vec<u8>>

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 8]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        req.set_att_key_id(att_key_id.clone());
+        req.set_b_pub_key_id(true);
+        req.set_buf_size(::std::mem::size_of::<sgx_sha256_hash_t>() as u64);
+        
+        let mut res = self.transact(req)?;
+
+        Ok(QuoteInfoEx { att_key_id,
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:189]: I don't think we need to return `att_key_id`. The user passed it in already. (After you remove this: unnecessary clone above)

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 9]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+        let mut req = Request_GetQuoteSizeExRequest::new();
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+        req.set_att_key_id(att_key_id);
+
+        let res = self.transact(req)?;
+
+        let quote_size : Option<u32> = if res.has_quote_size() { Some(res.get_quote_size()) } else { None };
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:213]: This seems a bit verbose. Is there no `take_` like you're using elsewhere?

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 10]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+    }
+
+    ///
+    /// Returns information needed by an intel SGX application to get a quote of one of its enclaves.
+    /// Similar functionality to sgx_init_quote_ex in page 165 at https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Developer_Reference_Linux_2.9.1_Open_Source.pdf
+    ///
+    /// @param[in] att_key_id   - a byte array with hidden type 'sgx_att_key_id_ext_t' returned by AesmClient::select_att_key_id or AesmKeyIds::select_algorithm_id or similar.
+    ///
+    /// @returns size of the quote buffer
+    pub fn get_quote_size_ex(&self, att_key_id: Vec<u8>) -> Result<Option<u32>> {
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:203]: I don't think this should return Result<Option<>>. Either you got an error, or you got a size.

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 11]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+
+    ///
+    /// Purpose of following functions is to stay as close as possible to protobuf interface.
+    /// for more usable API see AesmClient in rust-sgx/aesm-client/src/lib.rs
+    /// 
+    
+    ///
+    /// Returns number of keys stored in AESM service. This is used internally in get_supported_att_key_ids.
+    ///
+    pub fn get_supported_att_key_id_num(&self) -> Result<u32> {
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:121]: I think this should be inlined in `get_supported_att_key_ids` below

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 10 above)
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:203]: I think this should be inlined in `get_quote_size_ex` below

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 12]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+        req.set_report(report.clone());
+
+        // qe_report_info is of type sgx_ql_qe_report_info_t
+        // - nonce: sgx_quote_nonce_t
+        // - app_enclave_target_info: sgx_target_info_t
+        // - qe_report: sgx_report_t
+
+        let mut qe_report_info : Vec<u8> = Vec::new();
+
+        qe_report_info.extend(nonce.to_vec().clone());
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:245]: `.clone()` is redundant when calling `.to_vec()`. But even `to_vec` is unnecessary here.

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 13]
```diff
@@ -107,4 +109,150 @@ impl AesmClient {
...
+        qe_report_info.extend(quote_info.target_info);
+        qe_report_info.extend(report);
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:247]: If you build this first, you don't need to clone report above.

File: aesm-client/src/imp/windows.rs [hunk 14]
```diff
@@ -168,4 +168,20 @@ impl AesmClient {
         }
         return Ok(licence_token);
     }
+
+    pub fn get_supported_att_key_ids(&self) -> Result<AesmKeyIds> {
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs:172]: I prefer doing `#[cfg(not(windows))]` on the functions you add in src/lib.rs

File: aesm-client/src/lib.rs [hunk 15]
```diff
@@ -251,110 +319,54 @@ trait FromResponse: Sized {
...
-impl From<Request_GetLaunchTokenRequest> for Request {
-    fn from(r: Request_GetLaunchTokenRequest) -> Request {
-        let mut req = Request::new();
-        req.set_getLicTokenReq(r);
-        req
-    }
-}
+define_aesm_message!(Request_GetQuoteExRequest,  Response_GetQuoteExResponse,  set_getQuoteExReq,  has_getQuoteExRes,  take_getQuoteExRes);
+define_aesm_message!(Request_InitQuoteExRequest, Response_InitQuoteExResponse, set_initQuoteExReq, has_initQuoteExRes, take_initQuoteExRes);
+define_aesm_message!(Request_SelectAttKeyIDRequest, Response_SelectAttKeyIDResponse,  set_selectAttKeyIDReq, has_selectAttKeyIDRes, take_selectAttKeyIDRes);
```
Comment (jethrogb) [inline aesm-client/src/lib.rs:365]: Remove unused definition

File: aesm-client/tests/live_quote.rs (hunk 2 above)
Comment (AdrianCX) [reply aesm-client/tests/live_quote.rs:52]: I copied that comment - need to remove it, it's not relevant :)

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 13 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:247]: both extend and set_report move the value, so one will need to be a clone (unless I missed something):

```
244 |         qe_report_info.extend(report);
    |                               ------ value moved here

236 |         req.set_report(report);
    |                        ------ value moved here
```

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 12 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:245]: thanks!

File: aesm-client/src/lib.rs (hunk 15 above)
Comment (AdrianCX) [reply aesm-client/src/lib.rs:365]: fixed

File: aesm-client/src/imp/windows.rs (hunk 14 above)
Comment (AdrianCX) [reply aesm-client/src/imp/windows.rs:172]: thanks

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 10 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:203]: made it private and used only within module

File: aesm-client/src/lib.rs (hunk 5 above)
Comment (AdrianCX) [reply aesm-client/src/lib.rs:269]: removed, thanks

File: aesm-client/Cargo.toml (hunk 1 above)
Comment (AdrianCX) [reply aesm-client/Cargo.toml]: thanks

File: aesm-client/tests/live_quote.rs (hunk 3 above)
Comment (AdrianCX) [reply aesm-client/tests/live_quote.rs]: done

File: dcap-provider/src/lib.rs (hunk 4 above)
Comment (AdrianCX) [reply dcap-provider/src/lib.rs]: removed, thanks

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 10 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs]: changed, thanks

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 9 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs]: fixed, thanks

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 11 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:121]: inlined

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 7 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:138]: Reason for class is to have algorithm 'select' on data easily available and safer, Vec<Vec> might make it harder to use and possibly unsafer since we may allow user data in it

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 6 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:136]: fixed, thanks

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 8 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:188]: This is a copy to make it easier for 'get_quote_ex' to work.

'get_quote_ex' also needs a key, either we put it as parameter to the function or we store it for convenience in the 'QuoteInfoEx'

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 13 above)
Comment (jethrogb) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:232]: You shouldn't need to move it here since you're copying the data into another vector. `extend_from_slice` maybe?

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 7 above)
Comment (jethrogb) [reply aesm-client/src/imp/aesm_protobuf/mod.rs]: It's not safer if you require the user to split the result into elements?

Comment (jethrogb) [reply aesm-client/src/imp/aesm_protobuf/mod.rs]: I meant `Vec<Vec<u8>>` btw

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 8 above)
Comment (jethrogb) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:183]: Why clone here if the caller can just as easily clone

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 7 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs]: Fixed, thanks

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 8 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs]: fixed as you mentioned - caller is now responsible for clone

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 13 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs]: fixed via extend_from_slice

File: aesm-client/src/lib.rs [hunk 16]
```diff
@@ -98,6 +98,36 @@ impl QuoteType {
...
+
+}
+*/
+#[derive(Debug)]
+pub struct QuoteInfoEx {
+    pub target_info: Vec<u8>,
+
+    pub pub_key_id: Vec<u8>,
+
+    pub pub_key_id_size: u64,
```
Comment (jethrogb) [inline aesm-client/src/lib.rs:128]: This shouldn't be returned, it's `pub_key_id.len()`

File: aesm-client/src/imp/aesm_protobuf/mod.rs [hunk 17]
```diff
@@ -107,4 +107,146 @@ impl AesmClient {
...
+        const SGX_SHA256_SIZE : u64 = 32;
+        let mut req = Request_InitQuoteExRequest::new();
+
+        // FIXME: remove conditional compilation after resolving https://github.com/fortanix/rust-sgx/issues/31
+        #[cfg(not(target_env = "sgx"))]
+        req.set_timeout(REMOTE_AESM_TIMEOUT_US);
+
+        req.set_att_key_id(att_key_id);
+        req.set_b_pub_key_id(true);
+        req.set_buf_size(SGX_SHA256_SIZE);
```
Comment (jethrogb) [inline aesm-client/src/imp/aesm_protobuf/mod.rs:188]: Shouldn't you be calling this API twice, the first time to get this size?

File: aesm-client/src/lib.rs [hunk 18]
```diff
@@ -98,6 +98,36 @@ impl QuoteType {
...
+    keys: Vec<u8>,
+}
+
+impl AesmKeyIds {
+    pub fn keys(&self) -> &[u8] {
+        &self.keys
+    }
+
+}
+*/
```
Comment (jethrogb) [inline aesm-client/src/lib.rs:121]: Delete comment

File: aesm-client/src/lib.rs [hunk 19]
```diff
@@ -98,6 +98,36 @@ impl QuoteType {
...
+#[derive(Debug)]
+pub struct QuoteInfoEx {
+    pub target_info: Vec<u8>,
+
+    pub pub_key_id: Vec<u8>,
+
```
Comment (jethrogb) [inline aesm-client/src/lib.rs:127]: Nit: unnecessary whitespace between fields

File: aesm-client/tests/live_quote.rs [hunk 20]
```diff
@@ -29,6 +31,7 @@ fn live_quote() {
 
     let quoteinfo = client.init_quote().expect("init quote");
     let ti = Targetinfo::try_copy_from(quoteinfo.target_info()).unwrap();
+
```
Comment (jethrogb) [inline aesm-client/tests/live_quote.rs:34]: Delete unnecessary change

File: aesm-client/tests/live_quote.rs [hunk 21]
```diff
@@ -41,3 +44,61 @@ fn live_quote() {
...
+        .einittoken_provider(AesmClient::new())
+        .build();
+
+    let client = AesmClient::new();
+
+    let key_ids = client.get_supported_att_key_ids().unwrap();
+
+    // Print the keys for debugging
+    for i in key_ids.iter() {
+        let ptr = i.as_ptr() as *const sgx_att_key_id_ext_t;
```
Comment (jethrogb) [inline aesm-client/tests/live_quote.rs:73]: If you only use `sgx_att_key_id_ext_t` for debugging, then you might as well not introduce the dependency

File: aesm-client/tests/live_quote.rs [hunk 22]
```diff
@@ -41,3 +44,61 @@ fn live_quote() {
...
+                     (*ptr).base.mrsigner.split_at(16), (*ptr).base.prod_id, (*ptr).base.extended_prod_id, (*ptr).base.config_id.split_at(16), (*ptr).base.family_id, (*ptr).base.algorithm_id);
+        }
+    }
+
+    // Select the ECDSA key that will be used later
+    let ecdsa_key_id = key_ids.into_iter().find(|id| SGX_QL_ALG_ECDSA_P256 == get_algorithm_id(id));
+
+    if let Some(key) = ecdsa_key_id {
+        match client.init_quote_ex(key.clone()) {
+            Err(Error::AesmCode(AesmError::UnexpectedError_1)) => {
```
Comment (jethrogb) [inline aesm-client/tests/live_quote.rs:86]: I don't think this case should be handled separately. This test shouldn't pass if it's not supported. Why not just do `client.init_quote_ex(key.clone()).unwrap()`?

File: aesm-client/tests/live_quote.rs [hunk 23]
```diff
@@ -41,3 +44,61 @@ fn live_quote() {
...
+                return;
+            },
+            Err(error) => { panic!("Error in init_quote_ecdsa: {:?}", error); },
+            Ok(quote_info) => {
+                println!("QuoteInfoEx: {:x?}", quote_info);
+
+                let ti = Targetinfo::try_copy_from(&quote_info.target_info).unwrap();
+                let report = report_test::report(&ti, &mut device).unwrap();
+                
+                let res = client.get_quote_ex(key, report.as_ref().to_owned(), quote_info, &NONCE);
```
Comment (jethrogb) [inline aesm-client/tests/live_quote.rs:99]: Should unwrap

File: aesm-client/src/lib.rs (hunk 16 above)
Comment (AdrianCX) [reply aesm-client/src/lib.rs:128]: thanks, fixed

File: aesm-client/src/imp/aesm_protobuf/mod.rs (hunk 17 above)
Comment (AdrianCX) [reply aesm-client/src/imp/aesm_protobuf/mod.rs:188]: That also works, updated to do that.

File: aesm-client/src/lib.rs (hunk 18 above)
Comment (AdrianCX) [reply aesm-client/src/lib.rs]: thanks

File: aesm-client/src/lib.rs (hunk 19 above)
Comment (AdrianCX) [reply aesm-client/src/lib.rs]: changed

File: aesm-client/tests/live_quote.rs (hunk 20 above)
Comment (AdrianCX) [reply aesm-client/tests/live_quote.rs]: removed row

File: aesm-client/tests/live_quote.rs (hunk 21 above)
Comment (AdrianCX) [reply aesm-client/tests/live_quote.rs]: removed it entirely

File: aesm-client/tests/live_quote.rs (hunk 22 above)
Comment (AdrianCX) [reply aesm-client/tests/live_quote.rs]: added a comment in case someone sees the failure and wants to debug, hopefully CI still passes.

File: aesm-client/tests/live_quote.rs (hunk 23 above)
Comment (AdrianCX) [reply aesm-client/tests/live_quote.rs]: thanks
